PluginBuilder/Controllers/ApiController.cs

```diff
@@ -106,24 +108,73 @@ public async Task<IActionResult> Download(
+
+        var result = await conn.QueryFirstOrDefaultAsync<(string? Url, long BuildId)>(
+            "SELECT b.build_info->>'url' AS Url, v.build_id AS BuildId " +
+            "FROM versions v " +
             "JOIN builds b ON b.plugin_slug = v.plugin_slug AND b.id = v.build_id " +
             "WHERE v.plugin_slug=@plugin_slug AND v.ver=@version",
             new
             {
                 plugin_slug = pluginSlug.ToString(),
                 version = version.VersionParts
             });
-        if (url is null)
+
+        if (result == default || result.Url is null)
             return NotFound();
 
-        await conn.InsertEvent("Download", new JObject
+        var clientIp = HttpContext.Connection.RemoteIpAddress?.ToString();
+
+        var lastEventType = await conn.GetLastEventTypeForIpAsync(clientIp, pluginSlug);
+        var eventInsertionResponse = await conn.InsertEvent("Download", new JObject
         {
             ["pluginSlug"] = pluginSlug.ToString(),
             ["version"] = version.ToString()
-        });
-        return Redirect(url);
+        }, clientIp, pluginSlug, result.BuildId);
+        if (eventInsertionResponse)
+        {
+            await conn.RecordPluginDownloadStatistics(pluginSlug, "install", version.VersionParts);
+        }
+        else if (lastEventType?.ToLower() == "uninstall")
+        {
+            await conn.RecordPluginDownloadStatistics(pluginSlug, "install", version.VersionParts);
+        }
+        return Redirect(result.Url);
+    }
+
+
+    [AllowAnonymous]
+    [HttpGet("plugins/{pluginSlug}/versions/{version}/uninstall")]
+    public async Task<IActionResult> Uninstall(
+    [ModelBinder(typeof(PluginSlugModelBinder))] PluginSlug pluginSlug,
+    [ModelBinder(typeof(PluginVersionModelBinder))] PluginVersion version)
+    {
+        await using var conn = await ConnectionFactory.Open();
+        var buildId = await conn.QuerySingleOrDefaultAsync<long>(
+            "SELECT build_id FROM versions WHERE plugin_slug = @plugin_slug AND ver = @version",
+            new
+            {
+                plugin_slug = pluginSlug.ToString(),
+                version = version.VersionParts
+            });
+
+        var clientIp = HttpContext.Connection.RemoteIpAddress?.ToString();
+        var lastEventType = await conn.GetLastEventTypeForIpAsync(clientIp, pluginSlug);
+        var eventInsertionResponse = await conn.InsertEvent("Uninstall", new JObject
+        {
+            ["pluginSlug"] = pluginSlug.ToString(),
+            ["version"] = version.ToString()
+        }, clientIp, pluginSlug, buildId);
+
+        if (eventInsertionResponse)
+        {
+            await conn.RecordPluginDownloadStatistics(pluginSlug, "delete", version.VersionParts);
+        }
+        else if (lastEventType?.ToLower() == "download")
+        {
+            await conn.RecordPluginDownloadStatistics(pluginSlug, "delete", version.VersionParts);
```

Comment:
both branch same?

Reply:
I have merge the else with the original 'if' for both scenario